File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/SemanticTraversal.java
package com.foundgine.core.semantic;

import com.foundgine.core.abstractions.*;
import java.util.*;

public record SemanticTraversal(EntityId source, String name, EntityId target, List<RelationshipId> path) {
	public SemanticTraversal {
		Objects.requireNonNull(source);
		Objects.requireNonNull(name);
		Objects.requireNonNull(target);
		path = List.copyOf(path);
		if (path.isEmpty())
			throw new IllegalArgumentException("A semantic traversal must contain at least one relationship.");
	}

	public RelationshipId firstRelationship() {
		return path.get(0);
	}
}

File: src/csharp/Foundgine.Core/Semantic/SemanticTraversal.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic;

/// <summary>
/// A logical, caller-facing traversal that may span multiple physical/semantic
/// relationships. It is an open-intent alias, not a provider join definition.
/// The path is expanded into ordinary semantic relationships before
/// authorization and planning, so every hop retains its own semantics.
/// </summary>
public sealed record SemanticTraversal(
    EntityId Source,
    string Name,
    EntityId Target,
    IReadOnlyList<RelationshipId> Path)
{
    public RelationshipId FirstRelationship => Path[0];
}
